File: LeetCode/2816. Double a Number Represented as a Linked List.cpp
class Solution {
public:
    // 리스트 뒤집기 함수
    ListNode *reverseList(ListNode *head) {
        ListNode *prev = nullptr;
        ListNode *cur = head;
        while (cur != nullptr) {
            ListNode *next = cur->next;
            cur->next = prev;
            prev = cur;
            cur = next;
        }
        return prev;
    }

    // 리스트 두 배로 만들기 함수
    ListNode *doubleIt(ListNode *head) {
        if (head == nullptr) {
            return nullptr;
        }

        // 리스트 뒤집기
        head = reverseList(head);
        ListNode *cur = head;
        int carry = 0;

        // 노드를 순회하면서 값을 두 배로 만들기
        while (cur != nullptr) {
            int doubled_value = cur->val * 2 + carry;
            cur->val = doubled_value % 10;
            carry = doubled_value / 10;

            // 마지막 노드이고 캐리가 남아있다면 새 노드 추가
            if (cur->next == nullptr && carry) {
                cur->next = new ListNode(carry);
                break;
            }
            cur = cur->next;
        }

        // 다시 리스트를 뒤집어 원래 순서로 복구
        return reverseList(head);
    }
};

File: 2. LeetCode/2816. Double a Number Represented as a Linked List.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        cur = head
        while cur:
            next = cur.next
            cur.next = prev
            prev = cur
            cur = next
        return prev

    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None

        # 연결리스트 뒤집기
        head = self.reverseList(head)
        cur = head
        carry = 0

        # 각 노드의 값을 두 배로 만듭니다.
        while cur:
            doubled_value = cur.val * 2 + carry
            cur.val = doubled_value % 10
            carry = doubled_value // 10
            # carry가 남아 있는데 노드가 끝이면 하나 추가
            if not cur.next and carry:
                cur.next = ListNode(carry)
                break
            cur = cur.next

        # 다시 뒤집기
        return self.reverseList(head)


s = Solution()
print(s.doubleIt([1, 8, 9]))
